### PR TITLE
Fix #829 avoid getFilename() of null error

### LINKED ISSUE #829: Fatal error while analyzing anonymous class
- PHPMD version: 2.9.1
- PHP Version: 7.4.10
- Installation type: composer
- Operating System / Distribution & Version: Alpine Linux in Docker

### Current Behavior
```
# vendor/bin/phpmd phpmd-test.php ansi cleancode,codesize,controversial,design,naming,unusedcode
PHP Fatal error:  Uncaught Error: Call to a member function getFileName() on null in /data/vendor/phpmd/phpmd/src/main/php/PHPMD/AbstractNode.php:201
Stack trace:
#0 /data/vendor/phpmd/phpmd/src/main/php/PHPMD/AbstractNode.php(117): PHPMD\AbstractNode->getFileName()
#1 /data/vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php(232): PHPMD\AbstractNode->getFirstChildOfType('FormalParameter...')
#2 /data/vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php(98): PHPMD\Rule\CleanCode\UndefinedVariable->collectParameters(Object(PHPMD\Node\MethodNode))
#3 /data/vendor/phpmd/phpmd/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php(71): PHPMD\Rule\CleanCode\UndefinedVariable->collect(Object(PHPMD\Node\MethodNode))
#4 /data/vendor/phpmd/phpmd/src/main/php/PHPMD/RuleSet.php(255): PHPMD\Rule\CleanCode\UndefinedVariable->apply(Object(PHPMD\Node\MethodNode))
#5 /data/vendor/phpmd/phpmd/src/main/php/PHPMD/Parser.php(259): PHPMD\RuleSet->apply(Object(PHPMD\Node\MethodNode))
#6 /data/vendor/php in /data/vendor/phpmd/phpmd/src/main/php/PHPMD/AbstractNode.php on line 201
```

### Expected Behavior
Working properly without `Fatal error`.

### Steps To Reproduce:
Run `cleancode` rule:
```
<?php
class TestCase
{
    public function getAnonymousClass()
    {
        return new class {
            public function get(): string
            {
                return '';
            }
        };
    }
}
```

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

### PR DESCRIPTION
**Pull request by kylekatarnls**
Type: bugfix
Issue: Resolves #829
Breaking change: no

### CODE REVIEW

**Comment by kylekatarnls** [comment]
This is actually a quick fix, so the todo for an proper fix. But still it would avoid fatal errors that can occur when some violations are triggered from inside an anonymous class.

File: src/main/php/PHPMD/AbstractNode.php [hunk 1]
```diff
@@ -194,11 +195,17 @@ public function getEndLine()
...
     public function getFileName()
     {
-        return (string)$this->node->getCompilationUnit()->getFileName();
+        $compilationUnit = $this->node instanceof AbstractASTArtifact
+            ? $this->node->getCompilationUnit()
+            : null;
+
+        return $compilationUnit
+            ? (string)$compilationUnit->getFileName()
+            : null; // @TODO: Find the name from some parent node
```
Comment (tvbeek) [inline src/main/php/PHPMD/AbstractNode.php:208]: Maybe it is an idea to create a new issue for this todo. But for the rest it looks good for me.

**Review by tvbeek — APPROVED** [review]

File: src/main/php/PHPMD/AbstractNode.php (hunk 1 above)
Comment (ravage84) [reply src/main/php/PHPMD/AbstractNode.php:208]: Yes, please create an issue.

Comment (kylekatarnls) [inline src/main/php/PHPMD/AbstractNode.php:208]:
```suggestion
            : null; // @TODO: Find the name from some parent node https://github.com/phpmd/phpmd/issues/837
```

Comment (kylekatarnls) [reply src/main/php/PHPMD/AbstractNode.php]: Created.

**Review by tvbeek — APPROVED** [review]

File: src/main/php/PHPMD/AbstractNode.php [hunk 2]
```diff
@@ -194,11 +195,17 @@ public function getEndLine()
     /**
      * Returns the name of the declaring source file.
      *
-     * @return string
+     * @return string|null
      */
     public function getFileName()
     {
-        return (string)$this->node->getCompilationUnit()->getFileName();
+        $compilationUnit = $this->node instanceof AbstractASTArtifact
```
Comment (ravage84) [inline src/main/php/PHPMD/AbstractNode.php:202]: I still think ternary on multiple lines are bad programming practices...

File: src/main/php/PHPMD/AbstractNode.php [hunk 3]
```diff
@@ -194,11 +195,17 @@ public function getEndLine()
...
+            ? $this->node->getCompilationUnit()
+            : null;
+
+        return $compilationUnit
```
Comment (ravage84) [inline src/main/php/PHPMD/AbstractNode.php:206]: I still think ternary on multiple lines are bad programming practices...

**Review by ravage84 — APPROVED** [review]

File: src/main/php/PHPMD/AbstractNode.php (hunk 2 above)
Comment (kylekatarnls) [reply src/main/php/PHPMD/AbstractNode.php:202]: When using ternaries, it's highly recommended to split it in 2 lines so get a warning for uncovered branch:
```php
// 100% of coverage:
$a = true ? call_this() : call_that();

// 66% of coverage:
$a = true
  ? call_this()
  : call_that();
```

Some are developing tools to detect `call_that()` is never called in the first example above. But still for now the best way to detect it is to have multiple lines.

Comment (ravage84) [reply src/main/php/PHPMD/AbstractNode.php:202]: I wouldn't use ternary then...

**Pull request merged**